libcloud/compute/drivers/scaleway.py

```diff
@@ -0,0 +1,615 @@
+            'range': 'Intensive',
+            'arch': 'x86_64',
+        },
+    },
+    {
+        'id': 'X64-60GB',
+        'name': 'X64-60GB',
+        'ram': 61440,
+        'disk': 700,  # TODO: Update to _minimum_ required storage
+        'bandwidth': 1000,
+        'price': 0.18,
+        'extra': {
+            'cores': 10,
+            'monthly': 89.99,
+            'range': 'Intensive',
+            'arch': 'x86_64',
+        },
+    },
+    {
+        'id': 'X64-120GB',
+        'name': 'X64-120GB',
+        'ram': 122880,
+        'disk': 1000,  # TODO: Update to _minimum_ required storage
+        'bandwidth': 1000,
+        'price': 0.36,
+        'extra': {
+            'cores': 12,
+            'monthly': 179.99,
+            'range': 'Intensive',
+            'arch': 'x86_64',
+        },
+    },
+]
+
+# The API also doesn't give location info, so we provide it ourselves, instead.
+SCALEWAY_LOCATION_DATA = [
+    {'id': 'par1', 'name': 'Paris 1', 'country': 'FR'},
+    {'id': 'ams1', 'name': 'Amsterdam 1', 'country': 'NL'},
+]
+
+
+class ScalewayResponse(JsonResponse):
+    valid_response_codes = [httplib.OK, httplib.ACCEPTED,
+                            httplib.CREATED, httplib.NO_CONTENT]
+
+    def parse_error(self):
+        return super(ScalewayResponse, self).parse_error()['message']
+
+    def success(self):
+        return self.status in self.valid_response_codes
+
+
+class ScalewayConnection(ConnectionUserAndKey):
+    """
+    Connection class for the Scaleway driver.
+    """
+
+    host = SCALEWAY_API_HOSTS['default']
+    allow_insecure = False
+    responseCls = ScalewayResponse
+
+    def request(self, action, params=None, data=None, headers=None,
+                method='GET', raw=False, stream=False, region=None):
+        if region:
+            old_host = self.host
+            self.host = SCALEWAY_API_HOSTS[region.id
+                                           if isinstance(region, NodeLocation)
+                                           else region]
+            if not self.host == old_host:
+                self.connect()
+
+        return super(ScalewayConnection, self).request(action, params, data,
+                                                       headers, method, raw,
+                                                       stream)
+
+    def add_default_headers(self, headers):
+        """
+        Add headers that are necessary for every request
+        """
+        headers['X-Auth-Token'] = self.key
+        headers['Content-Type'] = 'application/json'
+        return headers
+
+
+def _to_lib_size(size):
+    return int(size / 1000 / 1000 / 1000)
+
+
+def _to_api_size(size):
+    return int(size * 1000 * 1000 * 1000)
+
+
+class ScalewayNodeDriver(NodeDriver):
+    """
+    Scaleway NodeDriver
+    """
+
+    type = Provider.SCALEWAY
+    connectionCls = ScalewayConnection
+    name = 'Scaleway'
+    website = 'https://www.scaleway.com/'
+
+    SNAPSHOT_STATE_MAP = {  # TODO map all states
```

Comment:
I see a few TODOs here. Is this something that you plan to do next? Or just a warning for the next person that something is missing here?

Reply:
Could have sworn I'd replied to this one. This particular TODO is retained from the original PR. I haven't been able to find any other states that it might actually report, so i'm not sure it's even something that _can_ be done, but I've opted to leave it just in case a someone else knows better and can fill it in.

As to the others, I'm planning to handle that soon, though hopefully in a separate PR rather than this one, as it will require a fair amount of back-and-forth with Scaleway to work out the details. Ideally, that PR would remove the hard-coded sizes entirely, replacing them with API calls to Scaleway instead.

Reply:
@danhunsaker In case you need any support from Scaleway about our APIs just give me sign :)

Reply:
@sieben Just need two things to resolve the remaining TODOs, at the moment:

One, what are all the valid snapshot states we might see from the API?

And two, will the root volume of a server always be listed as volume `0` in the API response? Or is there a better way to determine that?

Reply:
@sieben Can you please give an answer here? Thanks!